File: 238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] output = new int[nums.length];
        output[0] = 1;
        
        for (int i = 1; i < nums.length; i++) {
            output[i] = output[i - 1] * nums[i - 1];
        }
        int temp = 1;
        for (int k = nums.length - 1; k >= 0; k--) {
            output[k] = output[k] * temp;
            temp = temp * nums[k];
        }      
        return output;
    }
}

File: 238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        output = [1]
        product = 1
        
        for i in range(1, len(nums)):
            output.append(output[i - 1] * nums[i - 1])

        for i in range(len(nums) - 2, -1, -1):
            product *= nums[i + 1]
            output[i] *= product
        
        return output
